Option for arg/return type hints and correct typing for std::filesystem::path

timohl opened the pull request:
<!--
Title (above): please place [branch_name] at the beginning if you are targeting a branch other than master. *Do not target stable*.
It is recommended to use conventional commit format, see conventionalcommits.org, but not required.
-->
## Description

<!-- Include relevant issues or PRs here, describe what changed and why -->
This PR adds the option for defining `arg_name` and `return_name` in a custom `type_caster`, which allows having different python type hints for arguments and return value. To check if `arg_name` or `return_name` is available or `name` should be used as fallback is implemented using the template classes `as_arg_type` and `as_return_type`.

Being entirely optional, this should not impact any existing `type_caster`.
This type hint specialization system is applied to the `type_caster` for `std::filesystem::path`.
Here, `Union[os.PathLike, str, bytes]` is the argument type hint and `Path` is the return type hint.
This is more accurate than the previous usage of `os.PathLike` for both argument and return type hint.

Also, most classes in pybind11::typing now support this feature for nested types, e.g., `py::typing::List<std::filesystem::path>` becomes `List[Union[os.PathLike, str, bytes]]` in arguments and `List[Path]` in return types.

I have added unit tests to check for the new signatures of stl/filesystem in typing containers and also to verify that `name` is used as fallback for other types (tested for `std::vector<std::filesystem::path>`).


## Suggested changelog entry:

<!-- Fill in the below block with the expected RestructuredText entry. Delete if no entry needed;
     but do not delete header or rst block if an entry is needed! Will be collected via a script. -->

```rst
Added option for different arg/return type hints to type_caster.
Updated stl/filesystem to use correct arg/return type hints.
Updated pybind11::typing to use correct arg/return type hints for nested types.
```

<!-- If the upgrade guide needs updating, note that here too -->
## Possible TODOs
* [x] Update `Custom type casters` documentation
* [x] ~~Update type hints for Eigen (currently uses `np.ndarray` as arg type hint but also takes lists and tuples (should probably be `numpy.typing.ArrayLike` + maybe `typing.Annotated` for shape/dtype annotation~~ (going into follow-up PR)

## Tests for py::typing classes:
* [x] Test for `std::vector<T>` (fallback)
* [x] Test for `Tuple<T...>`
* [x] Test for `Tuple<T, ellipsis>`
* [x] Test for `Dict<K, V>`
* [x] Test for `List<T>`
* [x] Test for `Set<T>`
* [x] Test for `Iterable<T>`
* [x] Test for `Iterator<T>`
* [x] Test for `Callable<Return(Args...)>`
* [x] Test for `Callable<Return(ellipsis)>`
* [x] Test for `Union<T...>`
* [x] Test for `Optional<T>`
* [x] Test for `TypeGuard<T>`
* [x] Test for `TypeIs<T>`

I would love to get feedback on this!
Especially on the "Possible TODO" regarding Eigen. Should I implement that here in this PR, or should I open a separate PR later?

Currently, I am working on adding typing stubs to Open3D (isl-org/Open3D#6917) using `pybind11-stubgen`.
Applying mypy/pyright on existing example code showed that a lot of type checks failed, which could be fixed by this PR.

rwgk commented on include/pybind11/cast.h:
```diff
@@ -34,6 +34,39 @@ PYBIND11_WARNING_DISABLE_MSVC(4127)
+struct is_descr<const descr<N, Ts...>> : std::true_type {};
+
+// Use arg_name instead of name when available
+template <typename T, typename Enable = void>
```
For consistency with existing pybind11 code, could you please make this `typename SFINAE = void`?

We already have 38 uses of `SFINAE`, but no `Enable`.

timohl replied:
Done

rwgk commented on include/pybind11/typing.h:
```diff
@@ -131,6 +131,13 @@ struct handle_type_name<typing::Tuple<Types...>> {
     static constexpr auto name = const_name("tuple[")
                                  + ::pybind11::detail::concat(make_caster<Types>::name...)
                                  + const_name("]");
+    static constexpr auto arg_name
```
How many of the added `arg_name` and `return_name` are covered by unit tests?

timohl replied:
Currently just for 3 classes (+ one test for list of lists) but I can add tests for all.
I added an overview of implemented tests vs changes classes in my top comment (under Possible TODOs).
Should I keep adding all those tests to the filesystem test function or split them off into their own test function?

rwgk replied:
Hm ... up to you.

Splitting off sounds best, if that's not too much trouble.

rwgk reviewed (commented):
Nice!
 
> Update Custom type casters documentation

Could you please try to include that in this PR? — For awareness, please see [this comment](https://github.com/pybind/pybind11/pull/5416#discussion_r1813962686). Not sure if that changes anything for you. If not, completely fine.

> Update type hints for Eigen

Best in a follow-on PR.

timohl commented:
> > Update Custom type casters documentation
> 
> Could you please try to include that in this PR? — For awareness, please see [this comment](https://github.com/pybind/pybind11/pull/5416#discussion_r1813962686). Not sure if that changes anything for you. If not, completely fine.

Sure! I can also try to adopt some of the documentation updates from #3862 as you suggested in #5416 since I am already at it.

timohl commented:
While thinking about a test function for `TypeGuard` and `TypeIs`, I realized, that it probably should behave differently.
So, I changed it in [9ad445d](https://github.com/pybind/pybind11/pull/5450/commits/9ad445d93b55d1d3863fa10132a4f7e617811ef8) to always use the return type of the nested type when available.
Both are used only in return type hints and narrow down the type of the first argument.
Since return type hints are usually narrower than argument types, I think return type is more appropriate here.
(If any typing expert disagrees, please let me know)

Outstanding task: "Update Custom type casters documentation"
I will probably do that tomorrow.

rwgk commented:
> One CI check failed, but that is a dependency timeout (maybe you can rerun it?).

Done and it passed (very quickly).

I can look at the code changes only later (might be a day or two).

rwgk commented on tests/test_docs_advanced_cast_custom.cpp:
```diff
@@ -0,0 +1,86 @@
+    // Python -> C++: convert a `PyObject` into a `Point2D` and return false upon failure. The
+    // second argument indicates whether implicit conversions should be allowed.
+    bool load(handle src, bool) {
+        // Check if handle is valid Sequence of length 2
```
Similarly here ...

Could you please try to use the pybind11 APIs?

Compare with `array_caster::convert_elements()` in pybind11/stl.h:

Roughly (totally untested):

```
        auto l = py::reinterpret_borrow<sequence>(seq);                             
        if (l.size() != 2) {                                                 
            return false;                                                       
        }
        double xy[2];
        ix = 0;                                                                      
        for (auto it : l) {                                                     
            py::make_caster<double> conv;                                                    
            if (!conv.load(it, convert)) {                                      
                return false;                                                   
            }
            xy[ix++] = py::cast_op<double>(conv);                                                                  
        }                                                                       
        value.x = xy[0];
        value.y = xy[1];
        return true;                                                            
```

timohl replied:
I have changed it to use the pybind11 API.

timohl commented on tests/test_docs_advanced_cast_custom.cpp:
```diff
@@ -0,0 +1,86 @@
+        // Decrement references (the tuple now owns x an y)
+        Py_DECREF(x);
+        Py_DECREF(y);
+        return t;
```
I have changed it to use the pybind11 API.

timohl commented:
I have changed the example to only use the pybind11 API.
Additionally, I have changed the test for `py::typing` classes to only use pybind11 API for the caster.

The documentation text was also changed a little bit:
* Added recommendation to use pybind11 API to warning about Python C API
* Added note about `return_value_policy` and `convert` arguments referring to their respective sections in the documentation

rwgk commented on docs/advanced/cast/custom.rst:
```diff
@@ -38,56 +56,82 @@ type is explicitly allowed.
+        // arguments and return values.
+        // The signature of our negate function would then look like:
+        // `negate(Sequence[float]) -> tuple[float, float]`
+        static constexpr auto arg_name = const_name("Sequence[float]");
```
`Sequence[int|float]` or `Sequence[Union[int, float]]`?

timohl replied:
According to [PEP 484 – Type Hints](https://peps.python.org/pep-0484/#the-numeric-tower) and the [typing documentation](https://typing.readthedocs.io/en/latest/spec/special-types.html#special-cases-for-float-and-complex):

> when an argument is annotated as having type float, an argument of type int is acceptable

So, if we want a Sequence of floats and/or ints, I think, `Sequence[float]` is correct and shorter.
But I do not mind if you prefer to include `int`, as it is still correct, but just adds a little redundancy.

timohl replied:
Although, now I see that I also used `Union[float,int]` in the `RealNumber` caster for the tests in [tests/test_pytypes.cpp](https://github.com/pybind/pybind11/pull/5450/files#diff-20bf1cca35b241c87fbf4ced7a3b55fd2421389f6861d06f67415c32d8c6bbc6).
However, I think that is ok there to just test the behavior of `arg_name`/`return_name`.
If you can think of a more sensible test class, I am open.

rwgk replied:
> According to PEP 484 – Type Hints and the typing documentation:
>> when an argument is annotated as having type float, an argument of type int is acceptable

I wasn't aware of that, thanks for the explanation.

> However, I think that is ok there to just test the behavior of `arg_name`/`return_name`.

Could you maybe add a terse comment, that `int` isn't really needed, but appears for testing purposes? — For people who may use the test as a starting point for their work.

But looks good to me as is.

timohl replied:
I have added a comment

rwgk replied:
Perfect, thanks!

rwgk commented on docs/advanced/cast/custom.rst:
```diff
@@ -38,56 +56,82 @@ type is explicitly allowed.
+            }
+            // Check if each element is either a float or an int
+            for (auto item : seq) {
+                if (!py::isinstance<py::float_>(item) and !py::isinstance<py::int_>(item)) {
```
Oversight: `and` -> `&&`

timohl replied:
Fixed

rwgk reviewed (approved):
Awesome work, thanks a lot!

@cryos could you help with a second set of eyes? Is there anything that stands out to you when glancing through?

timohl commented:
Made a small correction in the `docs/advanced/cast/overview` section in [6ea4704](https://github.com/pybind/pybind11/pull/5450/commits/6ea4704d538a9744568d92be7dead2f384626c3c).
I have searched the docs and this should be the only mention of `os.PathLike`, `pathlib.Path` or `std::filesystem::path`.

rwgk commented:
Since there was no feedback from anyone else, I looked through the entire PR again. I found only one tiny oversight (aa21ab5294a406e602e468c14e0b3dd53c0c1341).

I also merged master.

Waiting for CI to finish. When I see it passing I'll merge.

HolyBlackCat commented:
`static constexpr auto return_name = const_name("Path");` seems to be wrong, `Path` isn't a valid type? I've reported https://github.com/pybind/pybind11/issues/5477
